src/cloud-cli/meltano/cloud/api/config.py

```diff
@@ -102,6 +120,92 @@ def config_path(self) -> Path:
 
         return self._config_path
 
+    @staticmethod
+    def decode_jwt(token: str) -> dict[str, t.Any]:
+        """Decode a JWT without verifying.
+
+        Args:
+            token: the jwt to decode
+
+        Returns:
+            The decoded jwt.
+        """
+        return jwt.decode(token, options={"verify_signature": False})
+
+    @property
+    def _trks_and_pids(self) -> t.List[str]:
+        """Get tenant resource keys and project ids from id token.
+
+        Returns:
+            List of trks and pids in the form '<trk>::<jpid>`
+
+        """
+        with suppress(jwt.DecodeError):
+            decoded = self.decode_jwt(self.id_token)  # type: ignore
+            trks_and_pids = decoded.get("custom:trk_and_pid")
+            if trks_and_pids:
+                return [perm.strip() for perm in trks_and_pids.split(",")]
+        return []
+
+    @cached_property
+    def tenant_resource_keys(self) -> set[str]:
+        """Get the tenant resource keys from the ID token.
+
+        Returns:
+            The tenant resource keys found in the ID token.
+
+        """
+        return {perm.split("::")[0] for perm in self._trks_and_pids}
+
+    @cached_property
+    def internal_project_ids(self) -> set[str]:
+        """Get the internal project IDs from the ID token.
+
+        Returns:
+            The internal project IDs found in the ID token.
+
+        """
+        return {perm.split("::")[1] for perm in self._trks_and_pids}
+
+    @property
+    def internal_project_id(self) -> str:
+        """Get the internal project ID.
+
+        Returns:
+            Internal project ID.
+
+        Raises:
+            NoMeltanoCloudProjectIDError: when ID token includes no project IDs.
+            MeltanoCloudProjectAmbiguityError: when ID token includes more
+                than one project ID.
+        """
+        if len(self.internal_project_ids) > 1:
+            raise MeltanoCloudProjectAmbiguityError()
```

Comment:
Here and for the related error `MeltanoCloudTenantAmbiguityError`, can add printed instructions and/or other mitigation on this error? Could instead be added to the docstring, but I think a clear error message to the user could go a long way.

I think the mitigation is for user to set `MELTAN_CLOUD_INTERNAL_PROJECT_ID` value. Not documented, that I know of, but I think you mentioned that this follows automatically from our env var auto-parsing logic.

Also, can we safely print the observed keys to the user, since I believe those are non-secret and would (worst case) give them an option to learn their key and self-solve the issue if unexpected?

Reply:
Good callout. I've added a more helpful error message.